compiler/rustc_trait_selection/src/solve/mod.rs

```diff
@@ -398,6 +418,85 @@ impl<'a, 'tcx> EvalCtxt<'a, 'tcx> {
+            let r = ecx.infcx.probe(|_| {
+                let (_, certainty) = ecx.evaluate_goal(goal.with(
+                    tcx,
+                    ty::Binder::dummy(ty::ProjectionPredicate {
+                        projection_ty: alias,
+                        term: other,
+                    }),
+                ))?;
+                ecx.make_canonical_response(certainty)
+            });
+            debug!("evaluate_normalizes_to(..) -> {:?}", r);
+            r
+        };
+
+        if goal.predicate.0.is_infer() || goal.predicate.1.is_infer() {
+            bug!(
+                "`AliasEq` goal with an infer var on lhs or rhs which should have been instantiated"
+            );
+        }
+
+        match (
+            goal.predicate.0.to_alias_term_no_opaque(tcx),
+            goal.predicate.1.to_alias_term_no_opaque(tcx),
+        ) {
+            (None, None) => bug!("`AliasEq` goal without an alias on either lhs or rhs"),
+            (Some(alias), None) => evaluate_normalizes_to(self, alias, goal.predicate.1),
+            (None, Some(alias)) => evaluate_normalizes_to(self, alias, goal.predicate.0),
+            (Some(alias_lhs), Some(alias_rhs)) => {
+                debug!("compute_alias_eq_goal: both sides are aliases");
+
+                let mut candidates = Vec::with_capacity(3);
+
+                // Evaluate all 3 potential candidates for the alias' being equal
+                candidates.push(evaluate_normalizes_to(self, alias_lhs, goal.predicate.1));
+                candidates.push(evaluate_normalizes_to(self, alias_rhs, goal.predicate.0));
+                candidates.push(self.infcx.probe(|_| {
+                    debug!("compute_alias_eq_goal: alias defids are equal, equating substs");
+                    let nested_goals = self.infcx.eq(goal.param_env, alias_lhs, alias_rhs)?;
+                    self.evaluate_all_and_make_canonical_response(nested_goals)
+                }));
+
+                debug!(?candidates);
+
+                let candidates = candidates.into_iter().flatten().collect::<Box<[_]>>();
+
+                if candidates.is_empty() {
+                    return Err(NoSolution);
+                }
+
+                let one = candidates[0];
+                if candidates[1..].iter().all(|resp| resp == &one) {
+                    return Ok(one);
+                }
+
+                if let Some(response) = candidates.iter().find(|response| {
+                    response.value.certainty == Certainty::Yes
+                        && response.has_no_inference_or_external_constraints()
+                }) {
+                    return Ok(response.clone());
+                }
+
+                let certainty =
+                    candidates.iter().fold(Certainty::AMBIGUOUS, |certainty, response| {
+                        certainty.unify_and(response.value.certainty)
+                    });
+                self.make_canonical_response(certainty)
```

Comment:
can you move that into a separate method? something like `fn try_merge_responses` or whatever?